agent/http.go

```diff
@@ -740,16 +748,32 @@ func setLastContact(resp http.ResponseWriter, last time.Duration) {
 	resp.Header().Set("X-Consul-LastContact", strconv.FormatUint(lastMsec, 10))
 }
 
-// setMeta is used to set the query response meta data
-func setMeta(resp http.ResponseWriter, m structs.QueryMetaCompat) {
+func setMetaWithoutETag(resp http.ResponseWriter, m structs.QueryMetaCompat) {
 	setIndex(resp, m.GetIndex())
 	setLastContact(resp, m.GetLastContact())
 	setKnownLeader(resp, m.GetKnownLeader())
 	setConsistency(resp, m.GetConsistencyLevel())
+	if m.GetEmptyCacheResult() {
```

Comment:
Yes, the behavior is the following globally:

1. if agent supports empty unmodified responses, it send it in the request (using ReturnEmptyResultOnUnmodified), it will do this in 2 cases: the API caller asked for it (either because it used `If-None-Match` header, either because it used `X-Consul-Unmodified-Empty` new X-Consul-Index for existing SDKs), or if it is an agent local cache (because we are sure we already have the value, so we don't need the new value when it is not modified)
2. When server receives a Query with new flag `ReturnEmptyResultOnUnmodified` and the parameters do match after timeout (aka Index or Hash has not been modified), instead of returning the serialized response, it sends an empty response with flag `EmptyCacheResult` so the agent on the order side can distinguish between Errors, empty Collections... I wanted initially to do it with a Error, but it was way more complicated and would have force to have both aligned. While this approach allow to implement this optimization per RPC call in a simpler way
3. When the agent get the reply, if `EmptyCacheResult`, it returns 304 for API calls, or do not update the cache value for caches